From 2ddd14e79a71d7ee73fce3dbd550a310f187e159 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Mon, 8 Sep 2014 22:52:56 -0400
Subject: [fs] aio: add missing smp_rmb() in read_events_ring

Message-id: <1410216777-18522-8-git-send-email-jmoyer@redhat.com>
Patchwork-id: 93352
O-Subject: [RHEL7 PATCH 7/8] aio: add missing smp_rmb() in read_events_ring
Bugzilla: 1131312
RH-Acked-by: Zach Brown <zab@redhat.com>

This is a straight backport of the following commit.  This was tested
using the libaio test harness, aio-stress, and xfstests aio tests.

This patch addresses bug 1131312.

  commit 2ff396be602f10b5eab8e73b24f20348fa2de159
  Author: Jeff Moyer <jmoyer@redhat.com>
  Date:   Tue Sep 2 13:17:00 2014 -0400

    aio: add missing smp_rmb() in read_events_ring

    We ran into a case on ppc64 running mariadb where io_getevents would
    return zeroed out I/O events.  After adding instrumentation, it became
    clear that there was some missing synchronization between reading the
    tail pointer and the events themselves.  This small patch fixes the
    problem in testing.

    Thanks to Zach for helping to look into this, and suggesting the fix.

    Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
    Signed-off-by: Benjamin LaHaise <bcrl@kvack.org>
    Cc: stable@vger.kernel.org

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/aio.c b/fs/aio.c
index 0b311a8..d6cf32c 100644
--- a/fs/aio.c
+++ b/fs/aio.c
@@ -1023,6 +1023,12 @@ static long aio_read_events_ring(struct kioctx *ctx,
  head = ring->head;
  kunmap_atomic(ring);
 
+ /*
+  * Ensure that once we've read the current tail pointer, that
+  * we also see the events that were stored up to the tail.
+  */
+ smp_rmb();
+
  pr_debug("h%u t%u m%u\n", head, ctx->tail, ctx->nr_events);
 
  if (head == ctx->tail)
-- 
1.7.1